io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -892,29 +892,37 @@ execAtomically !time !tid !tlbl !nextVid0 action0 k0 =
                                       tvarId tvar `Map.notMember` writtenOuter)
                                     writtenSeq
                          ++ writtenOuterSeq
-          -- Skip the orElse right hand and continue with the k continuation
+          -- Skip the right hand and continue with the k continuation
           go ctl' read written' writtenSeq' createdOuterSeq nextVid (k x)
 
-        OrElseRightFrame k writtenOuter writtenOuterSeq createdOuterSeq ctl' -> do
-          -- Commit the TVars written in this sub-transaction that are also
-          -- in the written set of the outer transaction
-          !_ <- traverse_ (\(SomeTVar tvar) -> commitTVar tvar)
-                          (Map.intersection written writtenOuter)
-          -- Merge the written set of the inner with the outer
-          let written'    = Map.union written writtenOuter
-              writtenSeq' = filter (\(SomeTVar tvar) ->
-                                      tvarId tvar `Map.notMember` writtenOuter)
-                                    writtenSeq
-                         ++ writtenOuterSeq
-              createdSeq' = createdSeq ++ createdOuterSeq
-          -- Continue with the k continuation
-          go ctl' read written' writtenSeq' createdSeq' nextVid (k x)
-
       ThrowStm e ->
+        {-# SCC "execAtomically.go.ThrowStm" #-}
+        case ctl of
+          AtomicallyFrame -> do
+            -- Revert all the TVar writes
+            !_ <- traverse_ (\(SomeTVar tvar) -> revertTVar tvar) written
+            k0 $ StmTxAborted [] (toException e)
+
+          BranchFrame (CatchStmA h) k writtenOuter writtenOuterSeq createdOuterSeq ctl' ->
+            {-# SCC "execAtomically.go.BranchFrame" #-} do
+            -- Revert all the TVar writes within this catch
+            !_ <- traverse_ (\(SomeTVar tvar) -> revertTVar tvar) written
+            -- Execute the catch handler with an empty written set
+            let ctl'' = BranchFrame EmptyStmA k writtenOuter writtenOuterSeq createdOuterSeq ctl'
+            go ctl'' read Map.empty [] [] nextVid (h e)
+            --
+          BranchFrame _ _k writtenOuter writtenOuterSeq createdOuterSeq ctl' ->
+            {-# SCC "execAtomically.go.BranchFrame" #-} do
+            -- Revert all the TVar writes within this branch
+            !_ <- traverse_ (\(SomeTVar tvar) -> revertTVar tvar) written
+            go ctl' read writtenOuter writtenOuterSeq createdOuterSeq nextVid (ThrowStm e)
+
+      CatchStm a h k ->
         {-# SCC "execAtomically.go.ThrowStm" #-} do
-        -- Revert all the TVar writes
-        !_ <- traverse_ (\(SomeTVar tvar) -> revertTVar tvar) written
-        k0 $ StmTxAborted [] (toException e)
+        -- Execute the left side in a new frame with an empty written set
```

Comment:
I would add a comment:
```suggestion
-- Execute the left side in a new frame with an empty written set.
-- Rollback `TVar`s written since catch handler was installed,
-- but preserve ones that were set prior to it, as specified in the
-- [stm](https://hackage.haskell.org/package/stm/docs/Control-Monad-STM.html#v:catchSTM) package.
```

Reply:
Could you add the comment.